src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/transformers/SnapshotTrackerTransformer.kt

```diff
@@ -0,0 +1,235 @@
+                        push(fieldName)
+                        loadNewCodeLocationId()
+                        // STACK: obj, obj, className, fieldName, codeLocation
+                        invokeStatic(Injections::updateSnapshotOnFieldAccess)
+                        // STACK: obj
+                        if (opcode == PUTFIELD) loadLocal(valueLocal)
+                        // STACK: obj, [value]
+                        visitFieldInsn(opcode, owner, fieldName, desc)
+                        // STACK: [<empty> | value]
+                    }
+                )
+            }
+
+            else -> {
+                visitFieldInsn(opcode, owner, fieldName, desc)
+            }
+        }
+    }
+
+    override fun visitInsn(opcode: Int) = adapter.run {
+        when (opcode) {
+            AALOAD, LALOAD, FALOAD, DALOAD, IALOAD, BALOAD, CALOAD, SALOAD -> {
+                invokeIfInTestingCode(
+                    original = {
+                        visitInsn(opcode)
+                    },
+                    code = {
+                        // STACK: array, index
+                        dup2()
+                        // STACK: array, index, array, index
+                        loadNewCodeLocationId()
+                        // STACK: array, index, array, index, codeLocation
+                        invokeStatic(Injections::updateSnapshotOnArrayElementAccess)
+                        // STACK: array, index
+                        visitInsn(opcode)
+                        // STACK: value
+                    }
+                )
+            }
+
+            AASTORE, IASTORE, FASTORE, BASTORE, CASTORE, SASTORE, LASTORE, DASTORE -> {
+                invokeIfInTestingCode(
+                    original = {
+                        visitInsn(opcode)
+                    },
+                    code = {
+                        val arrayElementType = getArrayElementType(opcode)
+                        val valueLocal = newLocal(arrayElementType) // we cannot use DUP as long/double require DUP2
+
+                        // STACK: array, index, value
+                        storeLocal(valueLocal)
+                        // STACK: array, index
+                        dup2()
+                        // STACK: array, index, array, index
+                        loadNewCodeLocationId()
+                        // STACK: array, index, array, index, codeLocation
+                        invokeStatic(Injections::updateSnapshotOnArrayElementAccess)
+                        // STACK: array, index
+                        loadLocal(valueLocal)
+                        // STACK: array, index, value
+                        visitInsn(opcode)
+                        // STACK: <empty>
+                    }
+                )
+            }
+
+            else -> {
+                visitInsn(opcode)
+            }
+        }
+    }
+
+    private fun isArray(type: Type): Boolean = type.sort == Type.ARRAY
```

Comment:
Please move `isArray` and `isPrimitive` into `TransformationUtils.kt` and add kdoc.

Reply:
done